src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs

```diff
@@ -0,0 +1,216 @@
+			_overnightSwapCache.Initialize(savedState.Select(OvernightSwapCalculation.Create));
+			
+			//start calculation
+			CalculateAndChargeSwaps();
+			
+			//TODO if server was down more that a day.. calc N days
+		}
+
+		public void CalculateAndChargeSwaps()
+		{
+			lock (LockObj)
+			{
+				_currentStartTimestamp = DateTime.UtcNow;
+
+				//read orders syncronously
+				var openOrders = _orderReader.GetActive();
+
+				//filter orders that are already calculated
+				var calculatedIds = _overnightSwapCache.GetAll()
+					.Where(x => x.Time > LastInvocationTime)
+					.SelectMany(x => x.OpenOrderIds);
+				var filteredOrders = openOrders.Where(x => !calculatedIds.Contains(x.Id));
+
+				//start calculation in a separate thread
+				_threadSwitcher.SwitchThread(async () =>
+				{
+					foreach (var accountOrders in filteredOrders.GroupBy(x => x.AccountId))
+					{
+						var clientId = accountOrders.First().ClientId;
+						MarginTradingAccount account;
+						try
+						{
+							account = _accountsCacheService.Get(clientId, accountOrders.Key);
+						}
+						catch (Exception ex)
+						{
+							await ProcessFailedOrders(accountOrders, accountOrders.Key, null, ex);
+							continue;
+						}
+
+						foreach (var ordersByInstrument in accountOrders.GroupBy(x => x.Instrument))
+						{
+							var firstOrder = ordersByInstrument.FirstOrDefault();
+							IAccountAssetPair accountAssetPair;
+							try
+							{
+								accountAssetPair = _accountAssetsCacheService.GetAccountAsset(
+									firstOrder?.TradingConditionId, firstOrder?.AccountAssetId, firstOrder?.Instrument);
+							}
+							catch (Exception ex)
+							{
+								await ProcessFailedOrders(ordersByInstrument, account.Id, ordersByInstrument.Key, ex);
+								continue;
+							}
+
+							foreach (OrderDirection direction in Enum.GetValues(typeof(OrderDirection)))
+							{
+								var orders = ordersByInstrument.Where(order => order.GetOrderType() == direction).ToList();
+								if (orders.Count == 0)
+									continue;
+								
+								try
+								{
+									await ProcessOrders(orders, ordersByInstrument.Key, account, accountAssetPair, direction);
+								}
+								catch (Exception ex)
+								{
+									await ProcessFailedOrders(orders, account.Id, ordersByInstrument.Key, ex);
+									continue;
+								}
+							}
+						}
+					}
+				});
+			}
+		}
+
+		/// <summary>
+		/// Calculate overnight swaps for account/instrument/direction order package.
+		/// </summary>
+		/// <param name="instrument"></param>
+		/// <param name="account"></param>
+		/// <param name="accountAssetPair"></param>
+		/// <param name="direction"></param>
+		/// <param name="orders"></param>
+		/// <returns></returns>
+		private async Task ProcessOrders(IReadOnlyList<Order> orders, string instrument, MarginTradingAccount account,
+				IAccountAssetPair accountAssetPair, OrderDirection direction)
```

Comment:
It is better to use domain interfaces, not DTO. E.g. MarginTradingAccount -> IMarginTradingAccount